hnn_core/lfp.py

```diff
@@ -0,0 +1,241 @@
+                #    b = position vector of electrode with respect to end
+                #
+                # we want to compute the length of projection of "a"
+                # "b".
+                # H = a.cos(theta) = a.dot(b) / |a|
+                a = end - start
+                b = exyz - end
+                H = a.dot(b) / dis
+
+                # total longitudinal distance from start of compartment
+                L = H + dis
+
+                # if a.dot(b) < 0, projection will fall on the
+                # compartment.
+                if H < 0:
+                    H = -H
+
+                # distance ^ 2 of electrode to end
+                r_sq = np.linalg.norm(exyz - end) ** 2
+
+                # phi
+                num = np.sqrt(H ** 2 + r_sq) - H
+                denom = np.sqrt(H ** 2 + r_sq) - L
+                phi = 1. / dis * np.log(num / denom)
+
+            # x10000 for units of microV : nA/(microm*(mS/cm)) -> microV
+            vres.append(10000.0 * phi / (4.0 * np.pi * sigma))
+        return vres
+
+    def LFPinit(self):
+        lsec = get_all_sections()
+        n_sections = len(lsec)
+
+        self.imem_ptrvec = h.PtrVector(n_sections)
+        self.imem_vec = h.Vector(n_sections)
+        for i, s in enumerate(lsec):
+            seg = s(0.5)
+            # for seg in s # so do not need to use segments...?
+            # more accurate to use segments and their neighbors
+            self.imem_ptrvec.pset(i, seg._ref_i_membrane_)
+
+        self.vres = self.transfer_resistance(self.coord, method=self.method)
+        self.lfp_t = h.Vector()
+        self.lfp_v = h.Vector()
+
+    def callback(self):
+        # print('In lfp callback - pc.id = ',self.pc.id(),' t=',self.pc.t(0))
+        self.imem_ptrvec.gather(self.imem_vec)
+
+        # verify sum i_membrane_ == stimulus
+        # s = pc.allreduce(imem_vec.sum(), 1)
+        # if rank == 0: print pc.t(0), s
+
+        # sum up the weighted i_membrane_. Result in vx
+        # rx.mulv(imem_vec, vx)
+
+        val = 0.0
+        for j in range(len(self.vres)):
+            val += self.imem_vec.x[j] * self.vres.x[j]
+
+        # append to Vector
+        self.lfp_t.append(self.pc.t(0))
+        self.lfp_v.append(val)
+
+
+if __name__ == '__main__':
+    from hnn_core.pyramidal import L5Pyr
+    from hnn_core.network_builder import load_custom_mechanisms
+
+    load_custom_mechanisms()
+    cell = L5Pyr()
+
+    h.load_file("stdgui.hoc")
+    h.cvode_active(1)
+
+    ns = h.NetStim()
+    ns.number = 10
+    ns.start = 100
+    ns.interval = 50.0
+    ns.noise = 0.  # deterministic
+
+    nc = h.NetCon(ns, cell.synapses['apicaltuft_ampa'])
+    nc.weight[0] = 0.001
+
+    h.tstop = 2000.0
+
+    for method in ['lsa', 'psa']:
+        elec = LFPElectrode([0, 100.0, 100.0], pc=h.ParallelContext(),
+                            method='psa')
+        elec.setup()
+        elec.LFPinit()
+        h.run()
```

Comment:
This could be factored into a unit test. If you read the reference paper, I think you might find some corner cases in which the LSA and PSA method give the exact same result ... (I suspect)

Reply:
Is this the sort of setup you're thinking should go in `test_lfp,py` for testing the code directly?

Reply:
yeah one test like that which checks that LSA and PSA give roughly equal results. I think they would be more equal the further you go away from the cell.

And then one where you could just look at the `vres` and see if it's computed correctly. I'm worried that [lines like this](https://github.com/jasmainak/hnn-core/blob/lfp/hnn_core/lfp.py#L105-L106) might be incorrect (see also [this figure](https://www.frontiersin.org/files/Articles/197337/fncom-10-00065-HTML/image_m/fncom-10-00065-g001.jpg)). You can give different positions and see if the `vres` is computed correctly -- e.g., if it's above a segment vs if it's outside the segment etc.